application/src/main/java/org/thingsboard/server/service/install/SqlDatabaseUpgradeService.java

```diff
@@ -628,29 +632,41 @@ public void upgradeDatabase(String fromVersion) throws Exception {
                         PageLink pageLink = new PageLink(1000);
                         PageData<TbPair<UUID, String>> pageData;
                         List<ListenableFuture<?>> futures = new ArrayList<>();
-                        Set<TenantId> tenants = new HashSet<>();
+                        Set<UUID> tenants = new HashSet<>();
                         do {
                             pageData = assetDao.getAllAssetTypes(pageLink);
                             for (TbPair<UUID, String> pair : pageData.getData()) {
                                 TenantId tenantId = new TenantId(pair.getFirst());
                                 String assetType = pair.getSecond();
-                                if (tenants.add(tenantId)) {
-                                    try {
-                                        futures.add(dbUpgradeExecutor.submit(() ->
-                                                assetProfileService.createDefaultAssetProfile(tenantId)));
-                                    } catch (Exception e) {}
+                                if (tenants.add(pair.getFirst())) {
+                                    futures.add(dbUpgradeExecutor.submit(() -> {
+                                        try {
+                                            assetProfileService.createDefaultAssetProfile(tenantId);
+                                        } catch (Exception e) {}
+                                    }));
                                 }
 
                                 if (!"default".equals(assetType)) {
-                                    try {
-                                        futures.add(dbUpgradeExecutor.submit(() ->
-                                                assetProfileService.findOrCreateAssetProfile(tenantId, assetType)));
-                                    } catch (Exception e) {}
+                                    futures.add(dbUpgradeExecutor.submit(() -> {
+                                        try {
+                                            assetProfileService.findOrCreateAssetProfile(tenantId, assetType);
+                                        } catch (Exception e) {}
+                                    }));
                                 }
                             }
                             pageLink = pageLink.nextPageLink();
                         } while (pageData.hasNext());
 
+                        List<UUID> tenantsWithoutProfiles = tenantRepository.getIdsNotIn(tenants);
```

Comment:
This request will not work for 50K tenants. Usually, max number of params in the "IN (?,?,?) stmt is less than 1024.

As I have already mentioned before in our discussion, kindly do the following:

1. Iterate over all tenants and create default profiles.
2. Iterate over the results of the assetDao.getAllAssetTypes and create remaining, non default profiles.

Reply:
done